### PR TITLE
[WIP] POC: add `reference` sub-task to suggest tables which can be used as Vitess reference tables

### PR DESCRIPTION
**Pull request by rohit-nayak-ps**
### Description

TL;DR: Add a new sub-task `reference ` to suggest tables which can be used as Vitess reference tables. Builds on top of `keys` and another new sub-command `schema` (which gets table row count estimates from the db or a specified file).

### Background

Vitess allows users to specify `reference` tables (https://vitess.io/docs/22.0/reference/vreplication/reference_tables/)
in a sharded keyspace which are defined in a`VSchema`. Lookup tables which are used commonly in joins and which are
small enough to be cached in all shards are candidates for it. Ideally these should also not be updated often.

`vt` already has support for ingesting query logs and extracting various useful information using `keys` and 
`trace`.

For identifying candidates for reference tables we would like these criteria to be met:
* Are relatively small compared to the _operational_ tables, i.e. have fewer rows.
* Have low write QPS: they don't get updated as often. This is important since there could be a lag in the reference 
  table materialization
* Are part of a lot of joins: setting them up as reference tables means fewer cross-shard joins

### Miscellaneous
* To be able to denote that a query has been called `n` times without needing it to be present those many times in 
  the query log, I added the ability to give this number using a `sql` `query comment` like `/*vt+ 
  VT_USAGE_COUNT=3413 */`. 

### Method

* Use `keys` to get all tables involved, including write and read qps on those tables and number of queries they 
  participate in.
* Use `schema` to get number of rows currently in the tables
* Use thresholds to determine conditions defined above: (thresholds hard coded for poc)
  * table row count < 1000 
  * writes to table < 1% of total writes (probably should be smaller?)
  * joins to table for reads > 5% of total reads 

### Notes
* If the target is already sharded in production, we can use the cross-shard queries to advice us
* Can we simulate the suggestions to see if cost will be reduced and by how much, if reference tables are used

#### vt schema [flags]
Flags are `vtParams` fields required to connect to database. It currently generates a json with just table names and
row counts. We may want to get foreign key constraints to get additional info on which tables are expected to be in 
joins.

### vt reference <query log file> <schema info json>
It first calls the `keys` internal function and then the `schema` one and uses simple logic based on thresholds to
generate json output with recommended tables like
```json
{
  "Tables": [
    {
      "Name": "city",
      "NumWrites": 45,
      "NumReads": 7844,
      "JoinCount": 15688,
      "Rows": 600
    },
    {
      "Name": "address",
      "NumWrites": 1,
      "NumReads": 7845,
      "JoinCount": 15689,
      "Rows": 603
    },
    {
      "Name": "country",
      "NumWrites": 47,
      "NumReads": 7845,
      "JoinCount": 7844,
      "Rows": 109
    },
    {
      "Name": "language",
      "NumWrites": 9,
      "NumReads": 5314,
      "JoinCount": 5313,
      "Rows": 6
    }
  ]
}
```

### sakila
Added the `sakila` schema for making it easier to test and generate queries for `reference`

### Refactoring
* `keys` refactored so that it can be called internally. Maybe we can think of a GRPC based server similar to 
  `vtctld` so that external tools can call them directly. Can then serve as a good interface for sub-commands that 
  use other sub-commands
* `keys` input format error messages updated

### CODE REVIEW

**Comment by rohit-nayak-ps** [comment]
For now we delay the suggestions and just add the table rows metric via https://github.com/vitessio/vt/pull/64
